File: backend/src/services/menu_access.py
"""ユーザーごとの表示メニュー（POG / 中央 / 地方 / 競輪）の判定。

**判定の唯一の正本。** DB にも FastAPI にも依存しない純関数だけを置く。
フロント側の写しは `frontend/src/lib/menuAccess.ts`（同じ規則をそのまま
実装している）。**片方だけ変えないこと。**
`backend/tests/test_menu_access.py` が両者の対応を固定している。

## 二段構えになっている理由

| 段 | 関数 | 何をするか |
|---|---|---|
| 保存する値 | `normalize_menu_flags` | 管理者が付けた ON/OFF を規則に合わせて整える |
| 見せる値 | `resolve_menu_access` | 保存値にロール等の事情を足して**実際の可視性**を出す |

分けてあるのは、**管理画面のチェックボックスの状態**（＝保存値）と
**その人に実際に見えるもの**（＝可視性）が一致しないためである。
admin は自分のフラグが OFF でも全部見える（下記）ので、両者を1つの値に
まとめると管理画面が「OFF なのに見えている」を表現できなくなる。

## 規則

1. 🔴 **POG が ON なら中央と地方も必ず ON**（ユーザー指示・2026-09-10）。
   POG は中央・地方の出走と賞金の上に乗っている機能で、順位表から
   `/races/{id}` `/chihou/races/{id}` へ直接リンクしている。POG だけ
   見せると、そのリンクの先が全部 404 になる。
   **保存時にも読み出し時にも掛ける**（正規化を保存側だけに置くと、
   DB を直接更新された行が規則を破ったまま配信される）。
2. **admin はフラグを無視して全部見える**（ユーザー指示・2026-09-10）。
   管理者が自分のフラグを切って締め出されるのを防ぐため。
3. **競輪は当面 admin 限定**（ユーザー指示・2026-09-10）。フラグは保存
   するが可視性には効かせない。`/keirin` 配下は netkeirin（ウマい車券）
   への入稿トリガーと入稿設定の編集を含み、外部サービスへ影響するため。
   開放するときは `KEIRIN_REQUIRES_ADMIN` を False にする（それだけで
   フラグが効くようになる。他は触らなくてよい）。
"""

from __future__ import annotations

from dataclasses import asdict, dataclass

#: 競輪を admin 限定に据え置くか。False にするとフラグ（`menu_keirin`）が効く。
#: ⚠️ False にする前に、`/keirin/settings` と入稿トリガーを admin 限定へ
#: 分離すること（現状は同じルート配下にある）。
KEIRIN_REQUIRES_ADMIN = True

#: メニューのキー。DB 列名は `menu_` を前置したもの。
MENU_KEYS: tuple[str, ...] = ("pog", "jra", "chihou", "keirin")


@dataclass(frozen=True)
class MenuFlags:
    """4 メニューの ON/OFF。保存値にも可視性にも同じ型を使う。"""

    pog: bool = False
    jra: bool = True
    chihou: bool = True
    keirin: bool = False

    def to_dict(self) -> dict[str, bool]:
        """`{"pog": ..., "jra": ...}` を返す。API レスポンス用。"""
        return asdict(self)


def normalize_menu_flags(flags: MenuFlags) -> MenuFlags:
    """保存する値を規則に合わせて整える。

    POG が ON なら中央・地方を ON へ引き上げる（規則 1）。
    それ以外は与えられた値をそのまま通す。
    """
    if flags.pog:
        return MenuFlags(pog=True, jra=True, chihou=True, keirin=flags.keirin)
    return flags


def resolve_menu_access(role: str, flags: MenuFlags) -> MenuFlags:
    """その人に**実際に見えるもの**を返す。

    Args:
        role: `keiba.users.role`（`admin` / `member`）。
        flags: DB に入っている保存値。

    Returns:
        ナビの表示とルートガードの両方に使う可視性。
    """
    if role == "admin":
        return MenuFlags(pog=True, jra=True, chihou=True, keirin=True)

    normalized = normalize_menu_flags(flags)
    if KEIRIN_REQUIRES_ADMIN:
        return MenuFlags(
            pog=normalized.pog,
            jra=normalized.jra,
            chihou=normalized.chihou,
            keirin=False,
        )
    return normalized

File: frontend/src/lib/menuAccess.ts
/**
 * ユーザーごとの表示メニュー（POG / 中央 / 地方 / 競輪）の判定。
 *
 * 🔴 **バックエンド `backend/src/services/menu_access.py` の写し。**
 * 規則を変えるときは両方を直す。`backend/tests/test_menu_access.py` が
 * 対応を固定しており、片方だけ変えると落ちる。
 *
 * 写しを置いている理由は、ナビの描画（サーバコンポーネント）とルートガード
 * （`proxy.ts` = Edge）がどちらもこの判定を要るのに、**Edge から FastAPI を
 * 毎リクエスト叩けない**ため。`proxy.ts` は Auth.js の JWT に載せた保存値から
 * ここで可視性を組み立てる。
 *
 * ## 規則
 *
 * 1. POG が ON なら中央と地方も必ず ON（POG は両者のレースへ直接リンクする）
 * 2. admin はフラグを無視して全部見える
 * 3. 競輪は当面 admin 限定（`KEIRIN_REQUIRES_ADMIN`）
 */

/** 競輪を admin 限定に据え置くか。false にするとフラグが効く。 */
export const KEIRIN_REQUIRES_ADMIN = true;

/** メニューのキー。表示順もこの順。 */
export const MENU_KEYS = ["pog", "jra", "chihou", "keirin"] as const;

export type MenuKey = (typeof MENU_KEYS)[number];

export type MenuFlags = {
  pog: boolean;
  jra: boolean;
  chihou: boolean;
  keirin: boolean;
};

/** 何も分からないとき（未ログイン・取得失敗）の値。全部隠す。 */
export const NO_MENU_ACCESS: MenuFlags = {
  pog: false,
  jra: false,
  chihou: false,
  keirin: false,
};

/** 管理画面の既定（新規ユーザーの初期値）。DB の server_default と揃える。 */
export const DEFAULT_MENU_FLAGS: MenuFlags = {
  pog: false,
  jra: true,
  chihou: true,
  keirin: false,
};

/**
 * 保存する値を規則に合わせて整える。
 *
 * POG が ON なら中央・地方を ON へ引き上げる。管理画面のチェックボックスも
 * これを通してから描くので、POG を入れた瞬間に中央・地方が点く。
 */
export function normalizeMenuFlags(flags: MenuFlags): MenuFlags {
  if (flags.pog) {
    return { pog: true, jra: true, chihou: true, keirin: flags.keirin };
  }
  return flags;
}

/**
 * その人に**実際に見えるもの**を返す。
 *
 * @param role `keiba.users.role`（`admin` / `member`）
 * @param flags DB に入っている保存値
 */
export function resolveMenuAccess(
  role: string | undefined,
  flags: MenuFlags,
): MenuFlags {
  if (role === "admin") {
    return { pog: true, jra: true, chihou: true, keirin: true };
  }
  const n = normalizeMenuFlags(flags);
  if (KEIRIN_REQUIRES_ADMIN) {
    return { pog: n.pog, jra: n.jra, chihou: n.chihou, keirin: false };
  }
  return n;
}

/**
 * ルートのパスがどのメニューに属するか。`null` は誰でも見てよいもの
 * （マイページ・利用規約・お問い合わせなど）。
 *
 * ⚠️ **前方一致の順序に意味がある。** `/chihou/results` は「地方」であって
 * 「中央の実績」ではないので、`/chihou` を `/results` より先に見る。
 */
export function menuOfPath(pathname: string): MenuKey | null {
  if (pathname === "/pog" || pathname.startsWith("/pog/")) return "pog";
  if (pathname === "/chihou" || pathname.startsWith("/chihou/")) return "chihou";
  if (pathname === "/keirin" || pathname.startsWith("/keirin/")) return "keirin";
  // 予想（/yoso）は中央のレースに紐づくので中央に含める。
  for (const prefix of ["/races", "/results", "/yoso"]) {
    if (pathname === prefix || pathname.startsWith(`${prefix}/`)) return "jra";
  }
  return null;
}

/**
 * ログイン直後・締め出し時に送る先。
 *
 * 🔴 **`/races` 決め打ちにしてはいけない。** 中央が OFF の人を `/races` へ
 * 送ると、ガードがまた別の場所へ送り返してリダイレクトが循環する。
 * 見えるものが 1 つも無ければマイページへ送る（そこは常に見える）。
 */
export function landingPath(access: MenuFlags, pogYear?: number | null): string {
  if (access.jra) return "/races";
  if (access.chihou) return "/chihou/races";
  if (access.pog) return pogYear ? `/pog/${pogYear}` : "/pog";
  if (access.keirin) return "/keirin";
  return "/my";
}
